executor/index_lookup_hash_join.go

```diff
@@ -0,0 +1,348 @@
+			indexRanges:       copiedRanges,
+			keyOff2IdxOff:     e.keyOff2IdxOff,
+			joiner:            e.joiner,
+			maxChunkSize:      e.maxChunkSize,
+			workerID:          workerID,
+			joinChkResourceCh: e.joinChkResourceCh,
+			joinResultCh:      e.joinResultCh,
+		},
+		matchPtrBytes: make([][]byte, 0, 8),
+	}
+	return iw
+}
+
+func (iw *innerHashWorker) run(ctx context.Context, wg *sync.WaitGroup) {
+	var task *lookUpJoinTask
+	ok, joinResult := iw.getNewJoinResult()
+	if !ok {
+		return
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case task, ok = <-iw.taskCh:
+		}
+		if !ok {
+			break
+		}
+		if task.buildError != nil {
+			joinResult.err = errors.Trace(task.buildError)
+			break
+		}
+		err := iw.handleTask(ctx, task, joinResult)
+		if err != nil {
+			joinResult.err = err
+			break
+		}
+	}
+	if joinResult.err != nil || (joinResult.chk != nil && joinResult.chk.NumRows() > 0) {
+		iw.joinResultCh <- joinResult
+	}
+}
+
+func (iw *innerHashWorker) getNewJoinResult() (bool, *indexLookUpResult) {
+	joinResult := &indexLookUpResult{
+		src: iw.joinChkResourceCh[iw.workerID],
+	}
+	ok := true
+	select {
+	case joinResult.chk, ok = <-iw.joinChkResourceCh[iw.workerID]:
+	}
+	return ok, joinResult
+}
+func (iw *innerHashWorker) handleTask(ctx context.Context, task *lookUpJoinTask, joinResult *indexLookUpResult) error {
+	dLookUpKeys, err := iw.constructDatumLookupKeys(task)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = iw.fetchInnerResults(ctx, task, dLookUpKeys)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	var ok bool
+	ok, joinResult = iw.join2Chunk(joinResult, task)
+	if !ok {
+		return errors.New("join2Chunk failed")
+	}
+	it := task.lookupMap.NewIterator()
+	for key, rowPtr := it.Next(); key != nil; key, rowPtr = it.Next() {
+		iw.matchPtrBytes = task.matchKeyMap.Get(key, iw.matchPtrBytes[:0])
+		if len(iw.matchPtrBytes) == 0 {
+			ptr := *(*uint32)(unsafe.Pointer(&rowPtr[0]))
+			misMatchedRow := task.outerResult.GetRow(int(ptr))
+			iw.joiner.onMissMatch(misMatchedRow, joinResult.chk)
+		}
+		if joinResult.chk.NumRows() == iw.maxChunkSize {
+			ok := true
+			iw.joinResultCh <- joinResult
+			ok, joinResult = iw.getNewJoinResult()
+			if !ok {
+				return nil
+			}
+		}
+	}
+
+	return nil
+
+}
+func (iw *innerHashWorker) join2Chunk(joinResult *indexLookUpResult, task *lookUpJoinTask) (ok bool, _ *indexLookUpResult) {
+
+	for i := 0; i < task.innerResult.NumChunks(); i++ {
+		curChk := task.innerResult.GetChunk(i)
+		iter := chunk.NewIterator4Chunk(curChk)
+		for innerRow := iter.Begin(); innerRow != iter.End(); innerRow = iter.Next() {
+			ok, joinResult = iw.joinMatchInnerRow2Chunk(innerRow, task, joinResult)
+			if !ok {
+				return false, joinResult
+			}
+
+		}
+	}
+	return true, joinResult
+}
+
+func (iw *innerHashWorker) joinMatchInnerRow2Chunk(innerRow chunk.Row, task *lookUpJoinTask,
+	joinResult *indexLookUpResult) (bool, *indexLookUpResult) {
```

Comment:
Why does it return a `bool` variable to check whether there's an error? Just return `error` is ok.